projects/igniteui-angular/src/lib/grids/hierarchical-grid/hierarchical-grid.component.ts

```diff
@@ -547,18 +547,23 @@ export class IgxHierarchicalGridComponent extends IgxHierarchicalGridBaseDirecti
                     index: this.dataView.indexOf(rowData)
                 };
             } else {
-                const rowID = this.primaryKey ? rowData.rowID : this.data.indexOf(rowData.rowID);
                 // child rows contain unique grids, hence should have unique templates
                 return {
                     $implicit: rowData,
-                    templateID: 'childRow-' + rowID,
+                    templateID: {
+                        type: 'childRow',
+                        rowID: rowData.rowID
+                    },
                     index: this.dataView.indexOf(rowData)
                 };
             }
         } else {
             return {
                 $implicit: this.isGhostRecord(rowData) || this.isAddRowRecord(rowData) ? rowData.recordRef : rowData,
-                templateID: 'dataRow',
+                templateID: {
+                    type: 'dataRow',
+                    rowID: null
```

Comment:
Also in general we do use igx template outlet to cache rows, but since the template outlet is a more generic directive and is not necessarily associated with rows it would be better for the property to be called just "id".